docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+* set of acceptable values
+* if it is required or if it is optional
+
+Well defined API allows automatically generated documentation and variable checks.
+
+**Free of outer variables**: only variables consisting the role API or defined locally are allowed.
+Any other variable (e.g. inventory variable or a variable set in a play) are forbidden.
+A role free of any outer variable:
+* can be shared across different Ansible solutions (improves modularity)
+* is easier to test as only the role variables need to be set (improves role based testing)
+* independent of variable name changes outside of a role (improves role independence)
+
+### Role variables
+
+#### Where the role variables live (with annotations?)
+
+All role variables live under `default/main.yml` file.
+The file is to be structured in the following way:
+```yaml
+# List of required undefined variables:
+#: name=r_openshift_node_deployment_type, required=true, description="Deployment type", choices=['origin', 'openshift-enterprise']
+#: name=r_openshift_node_service_type, required=true, description="System service type"
+#: name=r_openshift_node_image_tag, required=true, description="Node image tag to pull for containerized deployment"
+
+# List of optional undefined variables
+#: name=r_openshift_node_system_images_registry, description="System image registry to pull images from", required_when=["r_openshift_node_is_containerized | bool", "r_openshift_node_is_node_system_container | bool"], required_when=["r_openshift_node_is_containerized | bool", "r_openshift_node_is_openvswitch_system_container | bool", "r_openshift_node_use_openshift_sdn | bool"]
+
+#: description="Set to deploy containerized node"
+r_openshift_node_is_containerized: false
+#: description="Set to deploy a node over AH"
+r_openshift_node_is_atomic: false
+```
+
+The role variables can be broken down intro three categories:
+* required role variables but undefined: role variables that must be always defined but without a default value
+* optional role variables but undefined: role variables that are undefined by default but can be defined
+* role variables that are always defined with a default value: role variables that can be overwritten
+
+All three categories can be annotated with the following keywords:
+* name: optional annotation used for role variables that are undefined, used when a table of role variables and pre-checks are generated
+* description: short and accurate description of a role variable
+* required: set to `false` if not specified, if set to `true` the role variable is required (forced by generated pre-checks tasks)
+* choices: a list of acceptable values of a role variables
+* required_when: a list of conjuncts under which a role variable must be defined
+
+A table of role variables and a sequence of pre-checks are generated from the specified annotations.
+
+All variables under `var/main.yml` are considered local variables.
+
+#### We should check all relevant role variables are properly set
```

Comment:
Where defaults are allowed then place defaults values for variables.  Where variables are required then I like testing for those.  I'd prefer a failure message rather than dysfunctional cluster.